### ISSUE TITLE
Implement caching of OpenShift image

### ISSUE THREAD

**Issue by hferentschik** [minishift/minishift#143]
Using 'cluster up' to provision OpenShift (see issue #141), means that the OpenShift Docker image needs to be downloaded initially. This adds startup time, especially compared to the current approach. It is also a bottle neck for acceptance tests (see issue #135). 

To alleviate this, we can cache the OpenShift Docker image. We can export it via 'docker export' and import it when required prior to 'cluster up'. In this case you have the download cost of the image only once.

**Comment by jorgemoralespou** [comment]
Is this going to support working as a developer where you probably use latest?

**Comment by jimmidyson** [comment]
It has to. We will have to think about how to manage this file cache (evictions, etc) so minishift doesn't wastefully fill up disk, especially as images are pretty big.

**Comment by jorgemoralespou** [comment]
That's one of the reasons I decided just to support local docker, cause everytime I create a new machine I would need to pull down all the images. 

I don't know if this is possible but an idea would be to have the minishift client when you do up start a local "registry mirror" and have the docker in the machine use that mirror. Then, any external pull would be cached in the mirror. 
The main difficulty is how to implement this mirror/cache in a smart way so that will not boat disk space. This will ultimately introduce some need for image management on the client side, where the host operating system might not support docker. I don't know if skopeo could help in this task or not.
In any case, pruning of used images in stopped machines would result in a new pull when the machine is started, which at the end is a small drawback. So I would keep only one of latest and all the tagged images, that would require manual pruning.

**Comment by jorgemoralespou** [comment]
Some ideas here.
Is an iso needed once a machine has been created? If not, I would provide a minishift setting (I don't think there is settings right now, although I think there should) with:
- num_isos_to_keep: This would be the amount of isos to keep. If there's a new iso to download and this number is already at the limit, I would ask which one to delete, defaulting to the latest, and leave the user the decision.

I would also implement as setting the:
- check_for_new_versions
- always_update

And some other settings you'll be probably find useful with time.

**Comment by hferentschik** [comment]
Ok, I did some investigations into using [containers/image](https://github.com/containers/image) which is the underlying library for the [skopeo](https://github.com/projectatomic/skopeo) CLI. The idea would be to be able to back up all images using as target format OSTree or OCI. Exporting into this format is actually possible, but as it turns out, containers/image only supports exporting to OSTree and OCI, there is no way to import the images back into a Docker registry (go figure). See also issue [#155](https://github.com/containers/image/issues/155) of containers/image. 

On top of things there are several issues with consuming  containers/image as a dependency. The lesser of which is issue [269](https://github.com/containers/image/issues/269). The bigger issue is that it uses a library which uses C extensions and hence does not cross compile - see issue [#268](https://github.com/containers/image/issues/268).

So integrating of containers/image would require to solve some general problems first, followed by the actual ability to get from OSTree to OCI. 

The _dir_ transport works both ways, is , however, just a better _docker save/load_. 

For a full blown image cache, I think we should make the integration with containers/image work. We might need in this case lend a hand there to move this along. 

For the short term and for just caching the main Origin image, I will resort for now to _docker save/load_.

**Comment by jorgemoralespou** [comment]
Adding @runcom to the thread to see what he thinks, as he works on OCI and skopeo.

Antonio, as I told you in Berlin, our use case is to be able to save docker images on the host and share them between different instances of minishift, to avoid download time. 

Can you look into what @hferentschik wrote and give some advice? Also, if there's other/better ways that you can think of, they are more than welcome ;-)

Thanks,

**Comment by runcom** [comment]
> Ok, I did some investigations into using containers/image which is the underlying library for the skopeo CLI. The idea would be to be able to back up all images using as target format OSTree or OCI. Exporting into this format is actually possible, but as it turns out, containers/image only supports exporting to OSTree and OCI, there is no way to import the images back into a Docker registry (go figure). See also issue #155 of containers/image.

@hferentschik 

What do you mean by "containers/image only supports exporting to OSTree and OCI"? with containers/image you can _export_ to:

- ostree
- dir (plain directory containing layers and manifest)
- docker-tar
- docker-daemon
- docker (registry, effectively pushing an image see http://rhelblog.redhat.com/2017/05/11/skopeo-copy-to-the-rescue/)
- oci
- containers-storage (CRI-O storage backend)
- atomic (the Atomic registry)

and probably other will pop up in the near future. 
What I'm referring to, speaking about APIs in containers/image is an `ImageDestination` (grep for that in c/image).
You can use that interface to _export_ (or _copy_ if you will) to whatever you want listed above.

> , there is no way to import the images back into a Docker registry (go figure). See also issue #155 of containers/image.

yes, there's #155 in c/image but still as a start:
```
$ skopeo copy docker://busybox oci:busybox_oci
$ skopeo copy oci:busybox_oci docker://myinternalcachingregistry.awesome/busybox
```

> So integrating of containers/image would require to solve some general problems first, followed by the actual ability to get from OSTree to OCI.

Also, if something is missing in c/image I'd be very welcome to have someone implement missing features... (/cc @giuseppe as well)

> On top of things there are several issues with consuming containers/image as a dependency. The lesser of which is issue 269. The bigger issue is that it uses a library which uses C extensions and hence does not cross compile - see issue #268.

those aren't that hard to fix imo

@hferentschik @jorgemoralespou btw, it would be really awesome to have minishift use c/image - if you guys want to prioritize some stuff for out team (think about having OSTree to OCI) you'd be also welcome I guess. Let me know.

**Comment by jorgemoralespou** [comment]
And this is how collaboration makes things happen. 

@hferentschik your turn ;-)

Thanks Antonio. I think that using c/image in minishift would be awesome and would show all the power and synergies we have. Thanks again.

**Comment by hferentschik** [comment]
> What do you mean by "containers/image only supports exporting to OSTree and OCI"?

What I mean is I can export into this format, but my understanding is that I cannot do the other way around. Taking images from OSTree OCI and importing them into a Docker registry. My understanding after the discussion on https://github.com/containers/image/issues/155 is that with OCI it would be easiest to do, but still has some missing pieces. 

I understand there are other formats as well, but I am after a file based format which also shares image layers. 

> skopeo copy oci:busybox_oci docker://myinternalcachingregistry.awesome/busybox

I am a bit confused, so I can copy from OCI to a Docker registry? This is contrary to what we said in https://github.com/minishift/minishift/pull/155. There must be somewhere a misunderstanding.

> Also, if something is missing in c/image I'd be very welcome to have someone implement missing features.

Awesome, as said, I think it would be nice to build Minishift's caching on top of c/image. I agree, the build issues are probably not hard to fix. There are also some cross compilation issues (we really need something which works on Linux, OS X and Windows), but this can for sure sorted out. Question is, how long will it take?

> it would be really awesome to have minishift use c/image

agreed

I have another look at all the issues I was facing and see what's really important for us. Some of the compilation errors I could work around already. Stay tuned ;-)

**Comment by runcom** [comment]
> but this can for sure sorted out. Question is, how long will it take?

https://github.com/containers/image/issues/155#issuecomment-300747176 :smile:

**Comment by giuseppe** [comment]
@jorgemoralespou would it make sense to use system containers where possible?  For instance, we could use it for the master, node, etcd and openvswitch containers and system containers already support to be preinstalled and be added at the image creation time.

Adding @mrguitar to the discussion

**Comment by jorgemoralespou** [comment]
@giuseppe currently that's a different problem from the one discussed in this issue, and it's how the team is creating the golden images they provide and if it would make sense for them to use system containers and cri-o to remove the dependency on docker on the VM, but that should be a long term goal.

In the meantime, we need to support just caching images out of the VM to be able to share, save/restore images.

**Comment by hferentschik** [comment]
@runcom, @jorgemoralespou - fyi, I was able to work around the problems with containers/image. Now I am running into troubles with containers/storage. I get several errors when trying to cross compile containers/storage. 

@runcom, to which extend are containers/storage and containers/image supposed to be OS independent libraries? Are there any parts which for example require a specific OS? If so, I would be fine if certain features are not available on some OSes, as long as the libraries cross compile.

**Comment by hferentschik** [comment]
Merged via pull request #1017
